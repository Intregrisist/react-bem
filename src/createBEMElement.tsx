
import * as PT from 'prop-types'
import * as React from 'react'

import { getDisplayName } from './helpers'
import resolveBEMNode from './resolveBEMNode'
import { BEMElementClass, BEMElementProps } from './types'

/**
 * Wraps a class with BEM element functionality, receiving the BEM block name
 * via context and converting element and modifiers attributes into className
 * attributes.
 * @param ComponentClass The class to wrap with BEM element functionality.
 */
export default function createBEMElement(
	ComponentClass: BEMElementClass | React.ComponentClass,
) {

	return class Wrapped<P = {}, S = {}>
	extends (ComponentClass as BEMElementClass)<P & BEMElementProps, S> {

		static displayName = `BEMElement(${getDisplayName(ComponentClass)})`

		static contextTypes = {
			...ComponentClass.contextTypes,
			block: PT.string.isRequired,
		}

		render() {
			const {
				element,
				modifiers,
			} = this.props
			const rendered = super.render.call(this)
			return rendered && resolveBEMNode(rendered, {
				block: this.context.block,
				element: element as string,
				modifiers,
			})
		}
	}
}
