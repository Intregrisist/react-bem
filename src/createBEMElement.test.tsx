import { shallow } from 'enzyme'
import * as PT from 'prop-types'
import * as React from 'react'

import createBEMElement from './createBEMElement'

describe('createBEMElement', () => {

	it('extends input class with a displayName of "BEMElement(Bar)"', () => {
		class Bar extends React.Component {
			render() {
				return null
			}
		}
		const BarElement = createBEMElement(Bar)
		expect(BarElement.displayName).toEqual('BEMElement(Bar)')
	})

	it('extends input propTypes with a block validator', () => {
		class Bar extends React.Component {
			static propTypes = {
				bar: PT.string,
			}
			render() {
				return null
			}
		}
		const BarElement = createBEMElement(Bar)
		expect(BarElement.propTypes).toMatchSnapshot()
	})

	it('receives BEM block name via props', () => {
		class Bar extends React.Component {
			render() {
				expect((this.props as any).block).toBe('foo')
				return null
			}
		}
		const BarElement = createBEMElement(Bar)
		shallow(
			<BarElement block="foo" element="bar" />,
		)
	})

	it('merges modifiers prop with root node\'s modifiers', () => {
		class Bar extends React.Component {
			render() {
				return (
					<div modifiers="mod1" />
				)
			}
		}
		const BarElement = createBEMElement(Bar)
		expect(shallow(
			<BarElement block="foo" element="bar" modifiers="mod2" />,
		)).toMatchSnapshot()
	})
})
