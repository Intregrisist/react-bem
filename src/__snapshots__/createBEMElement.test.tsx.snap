// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBEMElement merges modifiers prop with root node's modifiers 1`] = `
<div
  className="foo__bar foo__bar--mod1 foo__bar--mod2"
/>
`;

exports[`createBEMElement receives BEM block name via context 1`] = `
Object {
  "bemBlock": "context-block",
}
`;
